pkg/devfile/parser/data/2.0.0/types.go

```diff
@@ -138,10 +153,40 @@ type Container struct {
 
 // Endpoint holds information about how an application is exposed
 type Endpoint struct {
-	Attributes    map[string]string `json:"attributes,omitempty"`
-	Configuration *Configuration    `json:"configuration,omitempty"`
-	Name          string            `json:"name"`
-	TargetPort    int32             `json:"targetPort"`
+
+	// Map of implementation-dependant string-based free-form attributes.
+	//
+	// Examples of Che-specific attributes:
+	// - cookiesAuthEnabled: "true" / "false",
+	// - type: "terminal" / "ide" / "ide-dev",
+	Attributes map[string]string `json:"attributes,omitempty"`
+
+	// Describes how the endpoint should be exposed on the network.
+	// - `public` means that the endpoint will be exposed on the public network, typically through a K8S ingress or an OpenShift route.
+	// - `internal` means that the endpoint will be exposed internally outside of the main workspace POD, typically by K8S services, to be consumed by other elements running on the same cloud internal network.
+	// - `none` means that the endpoint will not be exposed and will only be accessible inside the main workspace POD, on a local address.
+	//
+	// Default value is `public`
+	Exposure string `json:"exposure,omitempty"`
+	Name     string `json:"name"`
+
+	// Path of the endpoint URL
+	Path string `json:"path,omitempty"`
+
+	// Describes the application and transport protocols of the traffic that will go through this endpoint.
+	// - `http`: Endpoint will have `http` traffic, typically on a TCP connection. It will be automaticaly promoted to `https` when the `secure` field is set to `true`.
+	// - `https`: Endpoint will have `https` traffic, typically on a TCP connection.
+	// - `ws`: Endpoint will have `ws` traffic, typically on a TCP connection. It will be automaticaly promoted to `wss` when the `secure` field is set to `true`.
+	// - `wss`: Endpoint will have `wss` traffic, typically on a TCP connection.
+	// - `tcp`: Endpoint will have traffic on a TCP connection, without specifying an application protocol.
+	// - `udp`: Endpoint will have traffic on an UDP connection, without specifying an application protocol.
+	//
+	// Default value is `http`
+	Protocol string `json:"protocol,omitempty"`
+
+	// Describes whether the endpoint should be secured and protected by some authentication process
+	Secure     bool `json:"secure,omitempty"`
+	TargetPort int  `json:"targetPort,omitempty"`
```

Comment:
The `TargetPort` is required in the definition: https://devfile.github.io/devfile/_attachments/api-reference.html
The omitempty should be removed

Reply:
Also the type should be `int32`, since the HostPort and ContainerPort defined in Kube Core v1 are all type of `int32`. https://godoc.org/k8s.io/api/core/v1#ContainerPort

Reply:
@yangcao77 updated port to int32,
 https://golang.org/pkg/encoding/json/#Marshal these tags are used while marshaling the data, it is just there to skip writing targetPort while converting it to json/yaml
if you do not specify targetPort in your devfile, schema validation would fail as it is `required`. you can test yourself